### PR TITLE
Core/adding dofs interface

### PR DESCRIPTION
**Pull request by RiccardoRossi**
this PR adds the interface of dofs and of DofsArray to the model_part.

this is another step to allow working with Kratos matrices from python.

The next step will be to revive an old discussion about giving access to mass, stiffness and damping matrices so that they can be manipulated from python

### CODE REVIEW

**Comment by RiccardoRossi** [comment]
just to add that this will have to wait for #6096 before being merged (mostly because i wrongly derived one PR from the other, they are actually quite independent)

File: kratos/python/add_strategies_to_python.cpp
Comment (loumalouomega) [inline kratos/python/add_strategies_to_python.cpp]: The file changes are confusing. What happened here?

File: kratos/python/add_node_to_python.cpp
Comment (loumalouomega) [inline kratos/python/add_node_to_python.cpp]: Finally!!!, we should expose the construction of MPC to python now that dof are fully available in python. Will be very practical

**Review by loumalouomega — COMMENTED** [review]
Some comments

**Comment by armingeiser** [comment]
This does partially solve #5897.

File: kratos/python/add_node_to_python.cpp
Comment (RiccardoRossi) [reply kratos/python/add_node_to_python.cpp]: ok, but only after this PR is merged

File: kratos/python/add_strategies_to_python.cpp
Comment (RiccardoRossi) [inline kratos/python/add_strategies_to_python.cpp]: for whatever reason it formatted the file, so there were clashes all over

the only changes needed are these ones

**Review by RiccardoRossi — COMMENTED** [review]
corrected to remove formatting changes

File: kratos/python/add_strategies_to_python.cpp
Comment (RiccardoRossi) [reply kratos/python/add_strategies_to_python.cpp]: wrong formatting, removed

**Comment by RiccardoRossi** [comment]
@armingeiser i remember that some of you guys had an issue about exporting M, K, D matrices to python.

that was closed long ago, since no one stepped up for the implementation. do you happen to remember were the discussion was? i would like to resume it as a follow up of this

**Comment by armingeiser** [comment]
> @armingeiser i remember that some of you guys had an issue about exporting M, K, D matrices to python.
> 
> that was closed long ago, since no one stepped up for the implementation. do you happen to remember were the discussion was? i would like to resume it as a follow up of this

I don't remember to be honest, i can have a more detailed look tomorrow. Maybe it was related to the testing of the eigensystem solver of the EigenSolversApplication. I think that was solved finally by @oberbichler implementing the matrixmarket interface to the Kratos sparse matrices.

**Comment by oberbichler** [comment]
@armingeiser we found a solution to test them without implementing the MatrixMarket-IO

File: kratos/python/add_strategies_to_python.cpp
Comment (rubenzorrilla) [inline kratos/python/add_strategies_to_python.cpp]:
```suggestion
                    for (auto &r_dof : self)
```

File: kratos/python/add_strategies_to_python.cpp
Comment (rubenzorrilla) [inline kratos/python/add_strategies_to_python.cpp]:
```suggestion
                        values[counter++] = r_dof.GetSolutionStepValue();
```

File: kratos/python/add_strategies_to_python.cpp
Comment (rubenzorrilla) [inline kratos/python/add_strategies_to_python.cpp]:
```suggestion
                    for (auto &r_dof : self)
```

File: kratos/python/add_strategies_to_python.cpp
Comment (rubenzorrilla) [inline kratos/python/add_strategies_to_python.cpp]:
```suggestion
                        values[counter++] = r_dof.EquationId();
```

File: kratos/python/add_strategies_to_python.cpp
Comment (rubenzorrilla) [inline kratos/python/add_strategies_to_python.cpp]:
```suggestion
                    for (auto &r_dof : self)
```

File: kratos/python/add_strategies_to_python.cpp
Comment (rubenzorrilla) [inline kratos/python/add_strategies_to_python.cpp]:
```suggestion
                        r_dof.GetSolutionStepValue() = values[counter++];
```

File: kratos/python/add_strategies_to_python.cpp
Comment (rubenzorrilla) [inline kratos/python/add_strategies_to_python.cpp]:
```suggestion
                                    const Vector& values)
```

File: kratos/python/add_strategies_to_python.cpp
Comment (rubenzorrilla) [inline kratos/python/add_strategies_to_python.cpp]:
```suggestion
                                {
```

File: kratos/python/add_node_to_python.cpp
Comment (rubenzorrilla) [inline kratos/python/add_node_to_python.cpp]: At this point I think it makes sense to create an add_dof_to_python.cpp

**Review by rubenzorrilla — COMMENTED** [review]
Some minor comments.

File: kratos/python/add_node_to_python.cpp
Comment (loumalouomega) [reply kratos/python/add_node_to_python.cpp]: +1 to @rubenzorrilla I was thinking the same. I don't like when the python definition are mixed in an inconsistent way and I expend time to find the file where is defined

Comment (philbucher) [reply kratos/python/add_node_to_python.cpp]: +1, it is often quite tricky to find the py-exposure things

Comment (armingeiser) [reply kratos/python/add_node_to_python.cpp]: +1

Comment (RiccardoRossi) [reply kratos/python/add_node_to_python.cpp]: done

**Review by rubenzorrilla — APPROVED** [review]
No more concerns from my side

**Pull request merged**
